lib/tpm2_hmac_auth.c

```diff
@@ -31,9 +31,78 @@
 
 #include <tss2/tss2_sys.h>
 
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
 #include "log.h"
 #include "tpm2_error.h"
 #include "tpm2_nv_util.h"
+#include "tpm2_openssl.h"
+
+
+uint8_t *tpm2_get_cpHash(TSS2_SYS_CONTEXT *sapi_context, TPM2B_NAME entity_1_name,
+    TPM2B_NAME entity_2_name) {
+
+    SHA256_CTX sha256;
+    int is_success = SHA256_Init(&sha256);
+    if(!is_success) {
+        LOG_ERR("SHA256_Init failed when calculating HMAC");
+        return NULL;
+    }
+
+    uint8_t cmd_code[4];
+    TSS2_RC rval = Tss2_Sys_GetCommandCode(sapi_context, cmd_code );
+    if( rval != TPM2_RC_SUCCESS ) {
+        LOG_ERR("Failed to retrieve command code 0x%X", rval);
+        return NULL;
+    }
+    
+    is_success = SHA256_Update(&sha256, cmd_code, sizeof(cmd_code));
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+
+    is_success = SHA256_Update(&sha256, entity_1_name.name, entity_1_name.size);
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+    is_success = SHA256_Update(&sha256, entity_2_name.name, entity_2_name.size);
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+
+    size_t command_params_size;
+    const uint8_t *command_params;
+    rval = Tss2_Sys_GetCpBuffer( sapi_context, &command_params_size, &command_params);
+    if( rval != TPM2_RC_SUCCESS ) {
+        LOG_ERR("Failed to retrieve cp Buffer:%08X", rval);
+        return NULL;
+    }
+
+    is_success = SHA256_Update(&sha256, command_params, command_params_size);
+    if (!is_success) {
+        LOG_ERR ("SHA256_Update failed when calculating HMAC");
+        return NULL;
+    }
+
+    uint8_t *cpHash = (unsigned char*)malloc(SHA256_DIGEST_LENGTH);
```

Comment:
There's no need to cast here.

Reply:
Well if you use a cpp compiler it complains. Even so the cast should have been uint8_t * not unsigned char *.

Just drop the cast.

Reply:
fixed